contracts/JBNFTRewardDataSourceDelegate.sol

```diff
@@ -0,0 +1,401 @@
+  */
+  IJBToken721UriResolver private _tokenUriResolver;
+
+  /**
+    @notice
+    Contract Opensea-style metadata uri.  Learn more https://bit.ly/3NnBa9v.
+  */
+  string private _contractUri;
+
+  /**
+    @param projectId JBX project id this reward is associated with.
+    @param directory JBX directory.
+    @param maxSupply Total number of reward tokens to distribute.
+    @param minContribution Minimum contribution amount to be eligible for this reward.
+    @param _name The name of the token.
+    @param _symbol The symbol that the token should be represented by.
+    @param _uri Token base URI.
+    @param _tokenUriResolverAddress Custom uri resolver.
+    @param _contractMetadataUri Opensea-style contract metadata uri. 
+    @param _admin Set an alternate owner.
+  */
+  constructor(
+    uint256 projectId,
+    IJBDirectory directory,
+    uint256 maxSupply,
+    JBTokenAmount memory minContribution,
+    string memory _name,
+    string memory _symbol,
+    string memory _uri,
+    IJBToken721UriResolver _tokenUriResolverAddress,
+    string memory _contractMetadataUri,
+    address _admin
+  ) ERC721Rari(_name, _symbol) {
+    // JBX
+    _projectId = projectId;
+    _directory = directory;
+    _maxSupply = maxSupply;
+    _minContribution = minContribution;
+
+    // ERC721
+    _baseUri = _uri;
+    _tokenUriResolver = _tokenUriResolverAddress;
+    _contractUri = _contractMetadataUri;
+
+    if (_admin != address(0)) _transferOwnership(_admin);
+  }
+
+  //*********************************************************************//
+  // ------------------- IJBFundingCycleDataSource --------------------- //
+  //*********************************************************************//
+
+  function payParams(JBPayParamsData calldata _data) 
+    external 
+    view 
+    override 
+    returns (uint256 weight, string memory memo, IJBPayDelegate delegate) 
+  {
+    return (0, _data.memo, IJBPayDelegate(address(this)));
+  }
+
+  function redeemParams(JBRedeemParamsData calldata _data)
+    external
+    pure
+    override
+    returns (uint256 reclaimAmount, string memory memo, IJBRedemptionDelegate delegate)
+  {
+    return (0, _data.memo, IJBRedemptionDelegate(address(0)));
+  }
+
+  //*********************************************************************//
+  // ------------------------ IJBPayDelegate --------------------------- //
+  //*********************************************************************//
+
+  function didPay(JBDidPayData calldata _data) 
+    external 
+    override 
+  {
+    if (!_directory.isTerminalOf(_projectId, IJBPaymentTerminal(msg.sender))) revert INVALID_PAYMENT_EVENT();
+
+    if (_distributedSupply == _maxSupply) return;    
+
+    if (
+      _data.amount.value >= _minContribution.value &&
+      _data.amount.currency == _minContribution.currency
+    ) {
+      
+      _mint(_data.beneficiary, _nextTokenId);
+
+      ++_supply;
+      ++_nextTokenId;
+
+      ++_distributedSupply;
```

Comment:
iiuc, you can avoid having both _nextTokenId and distributedSupply (would save one sstore on every mint) -> isn't _nextTokenId == distributedSupply? Or is there a scenario where it wouldn't be the case that I'm missing?

Reply:
same with supply actually, looks like Rari burn doesn't change it

Reply:
+ those are capped to max supply -> can be unchecked for some more gas golfing